Docs site

keller-mark opened the pull request:
Fixes #796 
Towards #392 
Closes #730
Fixes #427
Fixes #421

Latest demo:  http://beta.vitessce.io

This adds a documentation site with Docusaurus https://docusaurus.io/

There is some confusing logic in `docs/src/pages/_App.js` to handle the URL query params and decide whether to render either: the config editor, an error message, or `<Vitessce/>`, which probably can be cleaned up.

To test this out, there are instructions in [`docs/README.md`](https://github.com/hubmapconsortium/vitessce/blob/keller-mark/docs/docs/README.md)

For deployment, I added `./push-docs.sh`. This builds the docs site twice, once with the `baseUrl` set to `/` (for vitessce.io) and once with `/vitessce-data/docs/{date}/{hash}/`. I couldn't figure out a way to only build once and serve with different `baseUrl` values without breaking the paths to stylesheets and scripts. The idea with the two builds would be to use the `/vitessce-data/docs/{date}/{hash}/` build for the staging demo, and once this is approved, move the `/` build to `vitessce.io` with the `./copy-prod.sh` script.

I moved the demo configs to `/src/demo/configs.js` and these are also used in the docs site by `/docs/src/pages/demos.js`

ilan-gold commented:
@keller-mark I think this may be relatively related to https://github.com/vitessce/vitessce/compare/keller-mark/webpack-5 but I have two questions:

1. Do we need to use the `umd` build [here](https://github.com/vitessce/vitessce/pull/810/files#diff-3e3c2a02fed2bb352e936f1d797f8a330c09479b65b5f7a2428cc141c3c3e40d), for example?   This has been stewing for a while but I think we should drop the `umd` builds since they take a while to build and have minimal usage (I don't know of anyone trying to use Vitessce in a script tag - seems a little outdated).
2. If we can get rid of `umd` would it be possible to hook up a `watch` on the main Vitessce build so we can develop the docs and the source code simultaneously?

I can help out with these if you think either makes sense, but wanted to get your take first.

keller-mark commented:
I would be hesitant to remove the UMD build right now because the ES build didn't seem to work when I tried to use it for this docs site. But I would be fine removing it from the CI test scripts to reduce the time. And yes, I think we can try to import directly from the source files.

That said, I think the bundling is messed up, and I am hoping Webpack 5 helps. For instance, I could not get the bundle to load in Observable, with either of the builds

keller-mark commented:
I see errors from nebula.gl / turf when i try the ES build:

```
../node_modules/@turf/difference/index.mjs
Can't import the named export 'diff' from non EcmaScript module (only default export is available)
 @ ../node_modules/nebula.gl/dist-es6/lib/deck-renderer/deck-drawer.js
 @ ../node_modules/nebula.gl/dist-es6/index.js
 @ ../dist/es/production/index.js
 @ ./src/pages/_ThemedVitessce.js
 @ ./src/pages/index.js
 @ ./.docusaurus/registry.js
 @ ./node_modules/@docusaurus/core/lib/client/exports/ComponentCreator.js
 @ ./.docusaurus/routes.js
 @ ./node_modules/@docusaurus/core/lib/client/clientEntry.js
 @ multi ./node_modules/react-dev-utils/webpackHotDevClient.js ./node_modules/@docusaurus/core/lib/client/clientEntry.js
```

ilan-gold commented:
Ok @keller-mark - I have run into these issues as well in other contexts so that all checks out.  I'll review this then.  I think I will do a small release to fix that windows problem/upgrade Viv (needed for a HuBMAP thing) and then we can focus on doing a release of this + webpack if it seems feasible.  I'd like to be able to do `watch` development with docs.

ilan-gold commented on docs/src/pages/_LiveEditor.js:
```diff
@@ -0,0 +1,100 @@
+        .addFile(baseUrl + '/dries.cells.json', 'cells', 'cells.json')
+        .addFile(baseUrl + '/dries.cell-sets.json', 'cell-sets', 'cell-sets.json');
+    // Add components.
+    const umap = vc.addView(dataset, "scatterplot", { mapping: "UMAP" });
+    const tsne = vc.addView(dataset, "scatterplot", { mapping: "t-SNE" });
```
I think we should note that these `mapping` names come from within the `json`

ilan-gold commented on docs/src/pages/_LiveEditor.js:
```diff
@@ -0,0 +1,100 @@
+    // Add components.
+    const umap = vc.addView(dataset, "scatterplot", { mapping: "UMAP" });
+    const tsne = vc.addView(dataset, "scatterplot", { mapping: "t-SNE" });
+    const cellSetsManager = vc.addView(dataset, "cellSets");
+    const cellSetSizesPlot = vc.addView(dataset, "cellSetSizes");
+    // Link the zoom levels of the two scatterplots.
+    vc.linkViews([umap, tsne], ["embeddingZoom"], [2.5]);
+    // Try un-commenting the line below!
+    //vc.linkViews([umap, tsne], ["embeddingTargetX", "embeddingTargetY"], [0, 0]);
```
Could we use the variables that contain these values instead even if we have to export them?

keller-mark replied:
I have made a new file at `src/app/constants.js` and have added the constants to the main bundle exports

ilan-gold reviewed (commented):
Leaving this as a first pass.  Will probably take a few more...it's a lot to go through.

One question - why can't we import directly from the Vitessce source code like
```javascript
import { Vitessce } from '../../src';
```
or the like?  I must be missing something here.

ilan-gold commented:
Ah I see, I get the same `turf` error when I try to import.  That is very frustrating, I really wish I understood what was happening here.

ilan-gold commented on README.md:
```diff
@@ -116,55 +48,72 @@ The development server will refresh the browser as you edit the code.
+- [vitessce-python](https://github.com/vitessce/vitessce-python): Python API and Jupyter widget.
+- [vitessce-r](https://github.com/vitessce/vitessce-r): R API and R htmlwidget.
 - [vitessce-data](https://github.com/hms-dbmi/vitessce-data): Scripts to generate sample data
 - [ome-tiff-tiler](https://github.com/hms-dbmi/ome-tiff-tiler): Docker container to generate tiles from OME-TIFF
```
(This being `ome-tiff-tiler`)

keller-mark commented on docs/docs/component-vitessce.md:
````diff
@@ -0,0 +1,171 @@
+}
+```
+
+<!--
````
I think there was a reason but I don't remember now

keller-mark commented on docs/docs/view-config-json.md:
```diff
@@ -0,0 +1,171 @@
+
+The layout property defines which visualization (and controller) components will be rendered, how they will be arranged on the screen, and optionally how they will map onto coordination scopes. Each layout object represents one "component" or "view", and must contain a component name `component`, width `w` and height `h`, and horizontal position `x` and vertical position `y`. Components are arran ...
+
+For more information about the components that are available, please visit the [Visualization Components](/docs/components/index.html) documentation page.
+
+For more information about the coordination types that are available, please visit the [Coordination Types](/docs/coordination-types/index.html) documentation page.
```
Yes, I have updated the auto-generated coordination type listing to include which component uses each. As for why the lookup exists, I think using a subset may help zustand to avoid unnecessary re-renders

ilan-gold commented on docs/docs/component-scatterplot.md:
```diff
@@ -0,0 +1,52 @@
+---
```
Should we do other components?  I can help out with this effort.

keller-mark replied:
Potentially but I think before we do that we should try out this example in a fresh create-react-app and confirm that it actually works - it has been a while since I tried that

ilan-gold commented on docs/docs/view-config-js.md:
```diff
@@ -0,0 +1,474 @@
+
+#### Parameters:
+- `dataset` (`VitessceConfigDataset`) - A dataset instance to be used for the data visualized in this view.
+- `component` (`string`) - A component name.
```
I think we should link back to the list of components here (and anywhere else there are "magic strings" that are actually constants.)

ilan-gold commented on docs/docs/view-config-js.md:
```diff
@@ -0,0 +1,474 @@
+
+const vc = new VitessceConfig("My config");
+const dataset = vc.addDataset("My dataset");
+const v1 = vc.addView(dataset, "spatial");
```
Relatedly, this component name (`spatial`) is exported as an object that contains the constants?  We should use that here, and note that it is exported.

ilan-gold commented on docs/docs/view-config-js.md:
```diff
@@ -0,0 +1,474 @@
+const vc = new VitessceConfig("My config");
+const dataset = vc.addDataset("My dataset");
+const v1 = vc.addView(dataset, "spatial");
+const v2 = vc.addView(dataset, "scatterplot", { mapping: "X_umap" });
```
Ditto about the component names.

ilan-gold commented on docs/docs/view-config-js.md:
```diff
@@ -0,0 +1,474 @@
+
+#### Parameters:
+- `views` (`VitessceConfigView[]`) - An array of view objects to coordinate together.
+- `cTypes` (`string[]`) - The coordination types on which to coordinate the views.
```
Again about linking back to docs/the exported variable that has the constants.

ilan-gold commented on docs/docs/view-config-js.md:
```diff
@@ -0,0 +1,474 @@
+const v2 = vc.addView(dataset, "spatial");
+vc.linkViews(
+    [v1, v2]
+    ["spatialZoom", "spatialTargetX", "spatialTargetY"]
```
These are exported constants too?

ilan-gold reviewed (commented):
Few more comments, but I think after this review, I will likely give a final review and then approve.

keller-mark commented:
From @sehilyi we should add screenshots of each component on http://localhost:3001/docs/components/index.html and then link to them from https://vitessce.github.io/vitessce-python/api_config.html#vitessce.constants.Component and the corresponding page for R

keller-mark commented:
Changes since previous demo:
- styles
- removed the "Live Editor" page under the documentation, replaced it with a dropdown to choose a view config syntax type on the App page

![Screen Shot 2021-01-08 at 12 01 31 AM](https://user-images.githubusercontent.com/7525285/103978108-b7ea1a80-5148-11eb-9fa9-f5db45238462.png)

- rearranged the URL and file upload inputs on the App page based on your comments and comments from Nils
- added a directory of markdown files at docs/docs/demos/ where optional longer descriptions about each demo can be added
- use the URL hash rather than the URL GET params
- added screenshots of each component to the "Visualization Components" page
- added a new page called "Data Types and File Types" under Features, where I moved some of the info about file types
- added documentation of the constants for strings `Component`, `CoordinationType`, `FileType`, `DataType`
- refactored the demo view config files: moved out of `src/app/api.js` and into `src/demo/` directory

ilan-gold commented on docs/src/pages/_use-hash-param.js:
```diff
@@ -0,0 +1,98 @@
+// Adapted from https://github.com/hejmsdz/use-hash-param/blob/3fd6818/src/index.js
```
Will this break existing view configs/urls?  If I understand it correctly, the docs will be deployed to the vitessce.io in which urls like `http://vitessce.io/?url=data:,.......` should work but with this new hash param, it seems like they won't quite work

ilan-gold replied:
For example: https://s3.amazonaws.com/vitessce-data/docs/2021-01-11/d6521c4/demos/index.html/?dataset=vanderbilt&theme=dark doesn't work (because it doesn't use hash?).  Maybe I have this wrong though.  I think we should support both?

keller-mark replied:
Ok. I have reverted to only supporting hash params.

ilan-gold replied:
Sorry, hash or not hash?  I thought the issue was hash.

keller-mark replied:
Sorry, I meant to write query params. https://s3.amazonaws.com/vitessce-data/docs/2021-01-13/5366fb3/index.html?dataset=vanderbilt

ilan-gold replied:
I think we also need to support `index.html/?` style as well because that is what the current website resolves to.  For example `vitessce.io?url=data:,....` -> `vitessce.io/?url=data:,....` so copy and pasting browser urls currently copies the later, but it currently does not work with this demo, I think.

keller-mark replied:
When the site is copied to `vitessce.io`, both `vitessce.io/index.html?url=data:,...` and `vitessce.io/?url=data:,...` should work because we have set the root index.html document

![Screen Shot 2021-01-14 at 2 18 46 PM](https://user-images.githubusercontent.com/7525285/104637912-6f17f180-5673-11eb-9f9d-bb089d491ff9.png)

So this is not really an issue on our end, but we can resolve it by using a different host name when accessing the S3 site.

http://vitessce-data.s3-website-us-east-1.amazonaws.com/docs/2021-01-14/5366fb3/?dataset=vanderbilt

The servers at `.s3-website-{region}.amazonaws.com` recognize `index.html` as the default page for subdirectories served over HTTP, while `s3.amazonaws.com` does not (this server works with object "keys" rather than file names, in which the concept of a directory does not really exist) https://docs.aws.amazon.com/AmazonS3/latest/dev/WebsiteEndpoints.html

ilan-gold commented on docs/src/pages/_use-hash-param.js:
```diff
@@ -0,0 +1,98 @@
+        if(unmounted) {
+            return;
+        }
+        console.log("handleHashChange");
```
Remove?

ilan-gold reviewed (commented):
Very concerned about the URL breaking issue but other than that looks pretty good.

keller-mark commented on docs/src/pages/_ControlledEditor.js:
```diff
@@ -0,0 +1,48 @@
+
+const DEFAULT_UPDATE_DEBOUNCE_DELAY = 100
+
+const ControlledEditor = ({ value, onChange, editorDidMount, ...props }) => {
```
I have added a schema validation step when the "load from editor" button is pressed

keller-mark commented:
The most recent commit adds a header to the current demo with a link to the beta version of the site:

![Screen Shot 2021-03-03 at 6 17 52 PM](https://user-images.githubusercontent.com/7525285/109885668-ea962800-7c4c-11eb-8199-1cd17be93624.png)

ilan-gold commented on docs/README.md:
````diff
@@ -0,0 +1,60 @@
+This documentation site depends on the Vitessce production library located in `../dist/`.
+
+```sh
+cd ..
+npm install
+npm run build-lib:prod
+npm link docs/node_modules/react
+cd docs
+```
+
+## Installation
+
+Install the docusaurus dependencies.
+
+```sh
+npm install
````
I think these steps are in the wrong order since `docs/node_modules/react` doesn't necessarily exist.  Maybe make a note about the first time you do these steps if they are right?

ilan-gold commented on docs/docs/components.mdx:
```diff
@@ -0,0 +1,107 @@
+
+### `layerController`
+
+The layer controller component provides an interface for manipulating the visualization layers displayed in the `spatial` component. The corresponding `spatial` component(s) can be specified through the [`spatialLayers`](/docs/coordination-types/index.html#spatialLayers) coordination type.
```
Need to merge to get the `spatialLayers` update as well

ilan-gold commented on src/demo/view-configs/hubmap.js:
```diff
@@ -0,0 +1,32 @@
+import { vapi } from '../utils';
+
+function getConfig() {
```
This example never loads the full genomic profiles for me (FF browser):
<img width="1464" alt="Screen Shot 2021-03-25 at 4 30 27 PM" src="https://user-images.githubusercontent.com/43999641/112539491-718b7b80-8d87-11eb-894e-7fd3ea34c035.png">

keller-mark replied:
I think the higlass server is currently down http://higlass.io/app and that gene annotations track loads from that server

ilan-gold reviewed (approved):
Does not seem like too much changed.  Once you merge in the most recent changes, I can help you do one last pass over the docs vis-a-vis documenting all the new additions and such, and then we can merge without my reviewing again.

will-moore commented:
Hi Mark,
 This seems like a reasonable place to share my recent experience with your docs (which look very nice 👍 ).
I don't know that this is useful feedback at-all, but I might as well share it while it's fresh, just in case...

I think I started at http://beta.vitessce.io/docs/components/index.html and read through that and the following Coordination Model and Types sections. The concepts are explained well, but I still had a hard time grokking all the terminology without any concrete examples. So the demos for the Coordination Types are great (especially when I realized you could `EDIT` to see the config). More of them would be fantastic: e.g. I couldn't understand what `embeddingType` coordination does.

Having then read through the View Configs section, I had an idea about the various parts of a config, but didn't really know where to start trying to put one together.
I wanted to show a raster image with polygons coordinated with a scatter plot. 
I started at http://beta.vitessce.io/app/index.html using the "Try an example" and spent a while trying to add to that. For some reason I missed the `EDIT` button the whole time (maybe it was off the bottom of my screen) so this made it harder to iteratively tweak and try a config.
I looked at the demos but none really matched what I wanted so in the end I took the config from http://beta.vitessce.io/index.html?dataset=codeluppi-2018 (once I'd found the `EDIT` button!) and served it locally so I could edit it in my text editor with version control and figure out what I could remove from it.

This allowed me to get just the components and data I needed, working together.

In summary, I think that more of the "Try a demo" links, especially to illustrate the concepts on the Coordination Model page would be helpful. E.g. "In this example the coordination object is:
```
"spatialZoom": {
  "A": -5.5
},
```
(still not sure what the "A" means)?
It also seems that you have various coordination objects like this one just for setting initial values, and not for coordinating the `spatialZoom` between components?
And it's probably just me, but I wished I'd seen the `EDIT` button sooner (maybe if it had been at the top)?

I also only just noticed that if I click `EDIT` on http://beta.vitessce.io/index.html?dataset=codeluppi-2018 then on the editor page I click `LOAD FROM EDITOR` (without having edited anything) I get a 400 error page.

Anyway, the docs are looking great, thanks!

After Ilan told me that I needed to use vitessce.io (not the beta) for OME-zarr, I went on to use my simplified config in https://github.com/will-moore/omero-vitessce, loading all the cells and ome-zarr data from OMERO, although for some reason the layers controller won't show the channel controllers when loading ome-zarr from OMERO or locally hosted. Only when I load it from remote s3 - more investigation needed...

<img width="1200" alt="Screenshot 2021-04-13 at 23 39 31" src="https://user-images.githubusercontent.com/900055/114631210-15e04e00-9cb4-11eb-8fa5-9a0d8d41bad7.png">

ilan-gold commented:
Hi @will-moore neither Mark nor I are currently actively working on this (Mark is on rotation in another lab and I am doing the Viv 3D work) but we definitely appreciate the feedback and will incorporate it once we are back to working on this more fully.

keller-mark commented:
This is now ready for re-review. It now uses the ESM build of Vitessce.

There are still some improvements to be made but I can make issues for them once this is merged:
- Moving beta.vitessce.io to vitessce.io and then the current vitessce.io to a new subdomain (legacy.vitessce.io or something - we can discuss)
- Adding "[Try a demo!](http://beta.vitessce.io/docs/coordination-types/#embeddingZoom)" links for the remaining coordination types 
- More tutorials
- Screenshots/thumbnails for demos

ilan-gold commented on docs/src/pages/_App.js:
```diff
@@ -0,0 +1,378 @@
+    return [upgradeSuccess, failureReason];
+  }
+
+  function handleEditorGo() {
+    let nextUrl;
+    if(loadFrom === 'editor') {
+      let nextConfig = pendingConfig;
+      if(syntaxType === "JS") {
+        nextConfig = window[JSON_TRANSLATION_KEY];
+      }
+      nextUrl = 'data:,' + encodeURIComponent(nextConfig);
+
+      const [valid, failureReason] = validateConfig(nextConfig);
+      if(!valid) {
+        setError(failureReason);
+        return;
+      }
+    } else if(loadFrom === 'url') {
+      nextUrl = pendingUrl;
+    } else if(loadFrom === 'file') {
+      nextUrl = 'data:,' + encodeURIComponent(pendingFileContents);
+    }
+    window.location.href = baseUrl + nextUrl;
+  }
```
Is there no better way to do this?  When I click the "load from editor" button, it does a full reload of the page - the page goes white for a second, then the homepage comes up, and finally the loaded view configuration.

I don't want to force a full refactor but it's kind of slow.

keller-mark replied:
To prevent breaking old `vitessce.io/?url=x` and `vitessce.io/?dataset=x` urls, we need to continue to support those query parameters here. I think the full reload is required because query parameters are designed to send parameters to the server (despite the lack of server-side usage here)

Related: limits on AWS S3 query parameter values cause 400 errors (when the `?url=x` value is too long):
> I also only just noticed that if I click EDIT on http://beta.vitessce.io/index.html?dataset=codeluppi-2018 then on the editor page I click LOAD FROM EDITOR (without having edited anything) I get a 400 error page.

One solution is to allow both query parameters and hash-based parameters:
- hash-based parameters like `vitessce.io/#url=x` will not need a full refresh and will not be sent to the server (so only subject to client-side limits)
- query parameter URLs like `vitessce.io/?url=x` will still work
   - maybe these should be deprecated and redirect to the hash-based equivalent?

keller-mark replied:
Hash param support added in https://github.com/vitessce/vitessce/pull/1088

ilan-gold replied:
Thanks!  I think this makes sense.  I think the limit on query parameters is another motivating use-case for this, so I'm happy to see it done.

ilan-gold reviewed (approved):
Went through the changes since last review.  Nothing jumps out to me besides the one comment.
